File: impala/impala/report/p10_report/p10_report.js
// Copyright (c) 2023, Impala and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["P10 Report"] = {
	"filters": [
		{
            "fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"width": "100px",
			"reqd": 1
        },
        {
            "fieldname": "year",
            "label": __("Year"),
            "fieldtype": "Link",
            "options":"Fiscal Year",
            "default": frappe.defaults.get_user_default('fiscal_year')
        },
{
            "fieldname": "employee",
            "label": __("Employee Number"),
            "fieldtype": "Link",
            "options": "Employee"
        },
                {
            'fieldname': 'employer_pin',
            'fieldtype': 'Data',
            'label': __('Employer PIN'),
            'hidden': 1,
        },
	],
     onload: function() {
                frappe.query_report.set_filter_value('employee', '');     
                frappe.query_report.set_filter_value('year', '');    
                var company = frappe.query_report.get_filter_value('company');       
                frappe.db.get_value("Company", company, "tax_id", function(value) {
                    console.log(value['tax_id'])
                    frappe.query_report.set_filter_value('employer_pin', value['tax_id'])
                });        
               
            }
};

File: impala/impala/report/p10_report/p10_report.py
import frappe

def execute(filters=None):
    columns = [
        # {
        #     "label": "EMPLOYEE",
        #     "fieldname": "employee",
        #     "fieldtype": "Link",
        #     "options": "Employee",
        #     "width": 150
        # },
        {
            "label": "EMPLOYEE'S PIN",
            "fieldname": "employee_pin",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": "EMPLOYEE`S NAME",
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "TOTAL EMOLUMENTS",
            "fieldname": "deductions",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": "TOTAL DEDUCTED",
            "fieldname": "earnings",
            "fieldtype": "Currency",
            "width": 120
        }
    ]

    data = fetch_data(filters)

    # Add a row for the totals at the beginning
    totals_row = calculate_totals(data)
    data.insert(0, totals_row)

    return columns, data

def fetch_data(filters):
    sql_query = f"""
        SELECT
            ss.employee AS employee,
            e.employee_pin AS employee_pin,
            e.employee_name AS employee_name,
            SUM(CASE WHEN sd.parentfield = 'deductions' THEN sd.amount ELSE 0 END) AS deductions,
            SUM(CASE WHEN sd.parentfield = 'earnings' THEN sd.amount ELSE 0 END) AS earnings
        FROM
            `tabSalary Slip` ss
        LEFT JOIN
            `tabSalary Detail` sd ON ss.name = sd.parent
        LEFT JOIN
            `tabEmployee` e ON ss.employee = e.name
        WHERE
            ss.start_date BETWEEN '{filters.get('year')}-01-01' AND '{filters.get('year')}-12-31'
            AND (ss.end_date IS NULL OR ss.end_date BETWEEN '{filters.get('year')}-01-01' AND '{filters.get('year')}-12-31')
    """

    if filters.get("employee"):
        sql_query += f" AND ss.employee = '{filters.get('employee')}'"

    sql_query += """
        GROUP BY
            ss.employee
    """

    data = frappe.db.sql(sql_query, as_dict=True)

    return data

def calculate_totals(data):
    totals = {
        "employee": "",
        "employee_pin": "",
        "employee_name": "Total B/F from prev List",
        "deductions": sum(item["deductions"] for item in data),
        "earnings": sum(item["earnings"] for item in data),
    }
    return totals
